From a2f501987dad996835b2245524ea02632698d597 Mon Sep 17 00:00:00 2001
From: Mikhail Khodjaiants
Date: Thu, 14 Jun 2012 11:14:28 -0400
Subject: Bug 382496 - Eclipse hangs on 'Terminate and Remove' in the CDT debugger

---
diff --git a/dsf-gdb/org.eclipse.cdt.dsf.gdb/src/org/eclipse/cdt/dsf/gdb/launching/GdbLaunch.java b/dsf-gdb/org.eclipse.cdt.dsf.gdb/src/org/eclipse/cdt/dsf/gdb/launching/GdbLaunch.java
index c65e8f9..1923234 100644
--- a/dsf-gdb/org.eclipse.cdt.dsf.gdb/src/org/eclipse/cdt/dsf/gdb/launching/GdbLaunch.java
+++ b/dsf-gdb/org.eclipse.cdt.dsf.gdb/src/org/eclipse/cdt/dsf/gdb/launching/GdbLaunch.java
@@ -83,12 +83,7 @@ public class GdbLaunch extends DsfLaunch
 
         // Create the dispatch queue to be used by debugger control and services 
         // that belong to this launch
-        final DefaultDsfExecutor dsfExecutor = new DefaultDsfExecutor(GdbLaunchDelegate.GDB_DEBUG_MODEL_ID);
-        // Bug 293109 comment 14
-        // We can use delayed task on the executor, but once it is shutdown, we don't want to execute them.
-        // For instance, in GDBBackend, we start a 30-second delayed task to check that GDB properly started.
-        dsfExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
-        
+        final DefaultDsfExecutor dsfExecutor = new DefaultDsfExecutor(GdbLaunchDelegate.GDB_DEBUG_MODEL_ID);        
         dsfExecutor.prestartCoreThread();
         fExecutor = dsfExecutor;
         fSession = DsfSession.startSession(fExecutor, GdbLaunchDelegate.GDB_DEBUG_MODEL_ID);
--
cgit v0.8.3.4-115-g1759
